MF-1425 - Support external UUIDs for Things and Channels

qs-wang opened the pull request:
MF-1425 Enhancement to supply an external UUID for Things and Channels.

### What does this do?
This is the same PR for [ MF-1425 - Enhancement for external UUID #1426 ](https://github.com/mainflux/mainflux/pull/1426), but tried to resolve the conflicts

### Which issue(s) does this PR fix/relate to?
Resolves #MF-1425 

### Notes
I cannot @aspnair. If anyone in Mainflux can review this, and may ask aspnair take a look as well.

aspnair committed:
MF-1425 - Test cases changes for SDK

MF-1425 Enhancement to supply an external UUID for Things and Channels.
These are the new testcases added for
- Things Service Testcases
- SDK Things and Channel Testcases

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>
Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

aspnair committed:
MF-1425 - Fixing Testcases

MF-1425 Enhancement to supply an external UUID for Things and Channels.
Because of the previous commits, the testcases were getting failed
because the way ID was modified.  This change is to make sure that
all testcases are revisited to get them fixed.

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>
Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

aspnair committed:
MF-1425 - Fixing review comments

Fixing the review comments provided.

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>
Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

aspnair committed:
MF-1425 - Fixing more review comments

Signed-off-by: Anand Sivaram Palassery <aspnair@gmail.com>
Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

qs-wang committed:
MF-1425 - Rebase mainflux master to resolve conflicts

MF-1425 Enhancement to supply an external UUID for Things and Channels.
Resolve conflicts.




Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

qs-wang committed:
MF-1425 - Fixing conflicts

MF-1425 Enhancement to supply an external UUID for Things and Channels.
Fixing the conflicts between aspnair master, and the mainflux master.

Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

dborovcanin commented on things/api/things/http/requests.go:
```diff
@@ -24,14 +25,29 @@ type createThingReq struct {
 	Metadata map[string]interface{} `json:"metadata,omitempty"`
 }
 
+// validateUUID: validates the externally supplied UUID as per standard
```
No need for ` validateUUID: ` prefix, `Validates the externally supplied UUID as per standard` is sufficient. You can also completely omit the comment since it says nothing more than the function name.

qs-wang replied:
I removed the comment, and agreed such comment adds almost no value.

dborovcanin commented on things/api/things/http/endpoint.go:
```diff
@@ -263,7 +265,7 @@ func createChannelEndpoint(svc things.Service) endpoint.Endpoint {
 		}
 
-		ch := things.Channel{Name: req.Name, Metadata: req.Metadata}
+		ch := things.Channel{Name: req.Name, ID: req.ID, Metadata: req.Metadata}
```
Please don't inline (just to be in sync with how we construct Thing):
```go
ch := things.Channel{
    Name:     req.Name, 
    ID:       req.ID,
    Metadata: req.Metadata,
}
```

qs-wang committed:
MF-1425 Fix the comment and code format per review comments

MF-1425 Enhancement to supply an external UUID for Things and Channels.
1. Remove un-valued comment for a private function
2. Format the code for better readibility

Signed-off-by: Q.S. Wang <wangqs_eclipse@yahoo.com>

manuio commented on api/things.yml:
```diff
@@ -636,7 +636,10 @@ components:
           type: string
           format: uuid
-          description: Thing unique identifier
+          description: Thing unique identifier.  This can be either
```
Looks like you are adding 2 white space after every dot.

qs-wang replied:
Family speaking it was not me, but fixed. ^_^

qs-wang committed:
MF-1425 Enhancement to supply an external UUID for Things and Channels.

Fix the format of the API document

Signed-off-by: Q.S. Wang <wangqs_eclipse@yahoo.com>

dborovcanin reviewed (dismissed):
Great job, thanks @aspnair for sending the initial implementation and @qs-wang for fixing the code.

drasko commented on things/mocks/commons.go:
```diff
@@ -83,3 +86,15 @@ func sortChannels(pm things.PageMetadata, chs []things.Channel) []things.Channel
 }
+
+func parseID(ID string) (id uint64) {
+	var slNum string
```
What does `slNum` stands for?

qs-wang replied:
I guess it is serialNum? I have changed it to 'serialNum'.

qs-wang committed:
MF-1425 Enhancement to supply an external UUID for Things and Channels.

Rename the variable to make it readible.

Signed-off-by: Q.s <wangqs_eclipse@yahoo.com>

dborovcanin committed:
Merge branch 'master' into master

drasko reviewed (approved):
LGTM
